// optimal/esm/InstanceBuilder.js
function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import Builder from './Builder';
import isObject from './isObject';
import instanceOf from './instanceOf';

var InstanceBuilder = function (_Builder) {
  _inheritsLoose(InstanceBuilder, _Builder);

  function InstanceBuilder(refClass, loose) {
    var _this;

    if (refClass === void 0) {
      refClass = null;
    }

    if (loose === void 0) {
      loose = false;
    }

    _this = _Builder.call(this, 'instance', null) || this;

    _defineProperty(_assertThisInitialized(_this), "loose", false);

    _defineProperty(_assertThisInitialized(_this), "refClass", null);

    _this.nullable();

    if ("production" !== process.env.NODE_ENV) {
      if (refClass) {
        _this.invariant(typeof refClass === 'function', 'A class reference is required.');
      }

      _this.loose = loose;
      _this.refClass = refClass;

      _this.addCheck(_this.checkInstance, refClass);
    }

    return _this;
  }

  var _proto = InstanceBuilder.prototype;

  _proto.checkInstance = function checkInstance(path, value, refClass) {
    if ("production" !== process.env.NODE_ENV) {
      if (refClass) {
        this.invariant(typeof refClass === 'function' && (value instanceof refClass || this.loose && isObject(value) && instanceOf(value, refClass)), "Must be an instance of \"" + this.typeAlias() + "\".", path);
      } else {
        this.invariant(isObject(value) && value.constructor !== Object, 'Must be a class instance.', path);
      }
    }
  };

  _proto.typeAlias = function typeAlias() {
    var refClass = this.refClass;
    return refClass ? refClass.name || refClass.constructor.name : 'class';
  };

  return InstanceBuilder;
}(Builder);

export { InstanceBuilder as default };
export function instance(refClass, loose) {
  if (refClass === void 0) {
    refClass = null;
  }

  return new InstanceBuilder(refClass, loose);
}
export function builder() {
  return instance(Builder);
}
export function regex() {
  return instance(RegExp);
}
export function date() {
  return instance(Date);
}
