// optimal/lib/index.js
"use strict";

exports.__esModule = true;
var _exportNames = {
  predicates: true,
  Builder: true,
  bool: true,
  custom: true,
  func: true,
  ArrayBuilder: true,
  array: true,
  InstanceBuilder: true,
  builder: true,
  instance: true,
  date: true,
  regex: true,
  NumberBuilder: true,
  number: true,
  ObjectBuilder: true,
  object: true,
  blueprint: true,
  ShapeBuilder: true,
  shape: true,
  StringBuilder: true,
  string: true,
  UnionBuilder: true,
  union: true
};
exports.default = exports.predicates = void 0;

var _optimal = _interopRequireDefault(require("./optimal"));

var _Builder = _interopRequireWildcard(require("./Builder"));

exports.Builder = _Builder.default;
exports.bool = _Builder.bool;
exports.custom = _Builder.custom;
exports.func = _Builder.func;

var _ArrayBuilder = _interopRequireWildcard(require("./ArrayBuilder"));

exports.ArrayBuilder = _ArrayBuilder.default;
exports.array = _ArrayBuilder.array;

var _InstanceBuilder = _interopRequireWildcard(require("./InstanceBuilder"));

exports.InstanceBuilder = _InstanceBuilder.default;
exports.builder = _InstanceBuilder.builder;
exports.instance = _InstanceBuilder.instance;
exports.date = _InstanceBuilder.date;
exports.regex = _InstanceBuilder.regex;

var _NumberBuilder = _interopRequireWildcard(require("./NumberBuilder"));

exports.NumberBuilder = _NumberBuilder.default;
exports.number = _NumberBuilder.number;

var _ObjectBuilder = _interopRequireWildcard(require("./ObjectBuilder"));

exports.ObjectBuilder = _ObjectBuilder.default;
exports.object = _ObjectBuilder.object;
exports.blueprint = _ObjectBuilder.blueprint;

var _ShapeBuilder = _interopRequireWildcard(require("./ShapeBuilder"));

exports.ShapeBuilder = _ShapeBuilder.default;
exports.shape = _ShapeBuilder.shape;

var _StringBuilder = _interopRequireWildcard(require("./StringBuilder"));

exports.StringBuilder = _StringBuilder.default;
exports.string = _StringBuilder.string;

var _UnionBuilder = _interopRequireWildcard(require("./UnionBuilder"));

exports.UnionBuilder = _UnionBuilder.default;
exports.union = _UnionBuilder.union;

var _types = require("./types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  exports[key] = _types[key];
});

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */
var predicates = {
  array: _ArrayBuilder.array,
  blueprint: _ObjectBuilder.blueprint,
  bool: _Builder.bool,
  builder: _InstanceBuilder.builder,
  custom: _Builder.custom,
  date: _InstanceBuilder.date,
  func: _Builder.func,
  instance: _InstanceBuilder.instance,
  number: _NumberBuilder.number,
  object: _ObjectBuilder.object,
  regex: _InstanceBuilder.regex,
  shape: _ShapeBuilder.shape,
  string: _StringBuilder.string,
  union: _UnionBuilder.union
};
exports.predicates = predicates;
var _default = _optimal.default;
exports.default = _default;
